src/server/pkg/deploy/cmds/cmds.go

```diff
@@ -28,15 +35,89 @@ import (
 	"github.com/pachyderm/pachyderm/src/server/pkg/obj"
 	"github.com/pachyderm/pachyderm/src/server/pkg/serde"
 
+	docker "github.com/fsouza/go-dockerclient"
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 )
 
-var defaultDashImage = "pachyderm/dash:1.9.0"
+var (
+	awsAccessKeyIDRE = regexp.MustCompile("^[A-Z0-9]{20}$")
+	awsSecretRE      = regexp.MustCompile("^[A-Za-z0-9/+=]{40}$")
+	awsRegionRE      = regexp.MustCompile("^[a-z]{2}(?:-gov)?-[a-z]+-[0-9]$")
+)
+
+const (
+	defaultDashImage           = "pachyderm/dash:1.9.0"
+	defaultJupyterhubHubImage  = "pachyderm/jupyterhub-pachyderm-hub:1.0.0"
+	defaultJupyterhubUserImage = "pachyderm/jupyterhub-pachyderm-user:1.0.0"
+	jupyterhubChartVersion     = "0.8.2"
+	configTemplate             = `
+apiVersion: v1
+contexts:
+- context:
+    cluster: "%s"
+    user: "%s"
+    namespace: "%s"
+  name: pachyderm-active-context
+current-context: pachyderm-active-context
+kind: Config
+`
+)
+
+func kubectl(stdin io.Reader, context *config.Context, args ...string) error {
+	tmpfile, err := ioutil.TempFile("", "transient-kube-config-*.yaml")
+	if err != nil {
+		return errors.Wrapf(err, "failed to create transient kube config")
+	}
+	defer os.Remove(tmpfile.Name())
+	tmpfile.Write([]byte(fmt.Sprintf(configTemplate, context.ClusterName, context.AuthInfo, context.Namespace)))
+	tmpfile.Close()
+
+	var environ []string = nil
+	if context != nil {
```

Comment:
This generates a kube config on-the-fly with the pachyderm context built on top of the base config. This is an improvement over what we had before because, previously, we just assumed the kube and pachyderm contexts pointed to the same cluster. But it still seems pretty hackey.

Ideally, we'd use the kube API rather than calling kubectl for deployment/undeployment, but it became way too cumbersome to try to reproduce the same functionality that kubectl does here.

That said, I'd love to hear alternative suggestions that might clean this up - it's the part I'm most unhappy with in this PR.

Reply:
Yeah that's tricky, we actually used to send this stuff directly over the kube api, but we moved to call kubectl directly because it proved to be so hard to replicate the functionality of kubectl, particularl with respect to auth. You could conceivably make this a little nicer by using structs from the kubeapi rather than templates and then serializing them to yaml when you want to give them to kubectl... not 100% sure if that's going to be nicer though. That's probably what I would do, but I'm not sure my sense of how to make deployment to kube nice hasn't always yielded the best results.

Reply:
Yeah using the proper kube objects makes sense, done